docs/data/data-grid/performance/performance.md

```diff
@@ -2,43 +2,51 @@
 
 <p class="description">Improve the performance of the DataGrid using the recommendations from this guide.</p>
 
-## Memoize inner components with `React.memo`
+## Extract static objects and memoize root props
```

Comment:
Is it fair to expect a visible performance improvement on scroll after this PR?

I tried to compare a build from this PR to the latest stable version with a 6x CPU slowdown:
v6.5.0: https://codesandbox.io/s/nostalgic-tdd-imricw?file=/demo.tsx

https://github.com/mui/mui-x/assets/13808724/ef3efc16-6195-4b80-b5fb-fc7d57a3a515

this PR: https://codesandbox.io/s/serene-frost-c0u0wh?file=/demo.tsx


https://github.com/mui/mui-x/assets/13808724/771c340b-c87b-4ef9-be5d-4f6c9e3286bb

Reply:
> Is it fair to expect a visible performance improvement on scroll after this PR?

Lol yes, that's the point of the PR.

> Note: both demos are being run in Codesandbox, which is using React in development mode, but I think it's OK for comparison like this

I'm not sure how it would affect it. The final profiles that I posted (in the top-level comment) were run in production mode, which is the only mode I trust for actual results. I generally use dev mode only to give me an idea of what to work on, in particular for rendering-heavy code where React adds a lot of validation & instrumentation.

I'll run more benchmarks to ensure the results are positive and I'll publish them early next week.

Have you tried comparing the [PR netlify deployment](https://deploy-preview-9037--material-ui-x.netlify.app/x/introduction/)? (which I assume is in production mode)

Reply:
> Have you tried comparing the [PR netlify deployment](https://deploy-preview-9037--material-ui-x.netlify.app/x/introduction/)? (which I assume is in production mode)

Yes, I did, but then thought this comparison might not be fair because the demos do not memoize the `slots` prop:

https://deploy-preview-9037--material-ui-x.netlify.app/x/react-data-grid/#commercial-version

https://github.com/mui/mui-x/assets/13808724/c95707b4-20ef-41e4-a925-14605260657b

---

https://mui.com/x/react-data-grid/#commercial-version

https://github.com/mui/mui-x/assets/13808724/0f4cc7a2-b38c-4bfc-868e-b69bbd7ae176

Reply:
> I'll run more benchmarks to ensure the results are positive and I'll publish them early next week.

Cool, thanks!

Reply:
Running proper benchmarks itself is a tricky task. For example, if you have the react-dev-tools extension installed in your browser, it's going to add a lot of `performance.mark()` calls _even when you're not using it_, so all the benchmarks run in that context are corrupted, or at least distorted (incidentally, it's also slowing down your day-to-day browsing if you have it enabled at all times). I generally benchmark results on a clean guest profile window, with as few applications open as possible, while avoiding to breathe too loudly.

Reply:
> Yes, I did, but then thought this comparison might not be fair because the demos do not memoize the slots prop:

Yeah and the demo you linked is actually incorrect for this PR, this PR adds `React.memo` and that demo also adds `React.memo`, so there is a double `memo`. I need to update the docs to fix that.

Reply:
> For example, if you have the react-dev-tools extension installed in your browser, it's going to add a lot of performance.mark() calls even when you're not using it, so all the benchmarks run in that context are corrupted, or at least distorted

My assumption was that both demos are similarly impacted by external factors (not related to this PR), and therefore performance gains should be noticeable anyway.
But maybe that assumption was incorrect.

Reply:
Oh so I just looked at the codesandbox code you posted for the comparison, the same comment I made earlier applies:

> Yeah and the demo you linked is actually incorrect for this PR, this PR adds React.memo and that demo also adds React.memo, so there is a double memo. I need to update the docs to fix that.

To be clear: the optimization that this PR does _is adding `React.memo` by default_. The difference is that this PR does it while also adding logic that will make things work as they do without `React.memo`. So the performance would be:

 - (master): slowest
 - (master) + `React.memo`: fastest
 - (PR) + `React.memo` + store updates logic: fast, but slightly less than above

Reply:
> To be clear: the optimization that this PR does is adding React.memo by default. The difference is that this PR does it while also adding logic that will make things work as they do without React.memo

Right, so for visible performance gains we should be comparing:
1. master - without `React.memo`
2. this PR - without `React.memo` (now included by default)

I quickly compared them (again in Codesandbox):
- v6.6.0: https://codesandbox.io/s/sweet-lalande-tzoi3s?file=/demo.tsx

https://github.com/mui/mui-x/assets/13808724/9249502b-951a-4106-8a7e-a4160c5a80e9

- this PR: https://codesandbox.io/s/summer-feather-giow8o?file=/demo.tsx

https://github.com/mui/mui-x/assets/13808724/29c49c86-cc6b-451a-901c-32ac8fb99b00

Reply:
Preliminary exploration, but from what I've seen the cell wrapper comes up a lot in the profiles. I haven't benchmarked again after splitting the cell & the wrapper, and the other refactoring, I'll test again.

The CPU profiles show a small improvement for this PR, but the experience while scrolling seems slightly better for master, though I don't trust myself much to scroll with the same intensity. I'll poke around some more.

Reply:
Found the issue. The [why-did-you-render whack-a-mole](https://github.com/mui/mui-x/pull/8998) PR was meant to go with this one. It ensures the `React.memo()` can properly abort rendering. I'll cherry-pick the commits and test again.

Reply:
I've added the performance report in a comment below, continue the discussion there.